### ISSUE TITLE
couch-to-postgres adapter. phase 2 tracking.

### ISSUE THREAD

**Issue by btbonval** [medic/medic-webapp#1299]
* HealthRight ANC use case, port these forms into views
  * extract form fields and generate form definitions: 8h #1300
  * precise forms: V, R, P, D, F. 4h #1300 
  * fake forms: ANCR, ANCV. 2h
    * dictionary translating form name to fake form #1309
  * manips: ANCR manip (other manips are trivial). 4h #1339 
    * legit registry?
    * current state? (still pregnant?)
    * maximum delivery date
    * estimated delivery date
    * delivery date (as comparison of maximum if no actual)
  * ANC variables as a table; trivial
    * visit schedule: not sure if this is used at present. 2h to port if it is used.
  * ANC monthly impact metrics #1339 
    * activity 4h
      * monthy report per form: total forms, total CHWs, % CHWs
      * deliveries expected per month, missing births per month
    * impact 8h
      * average EDD error
      * early registration and %
      * delivery breakdown; facility, unskilled, skilled-no-fac
      * deliveries against visit counts (0 visit, 1+ visit, 4+ visit)
        * above two with/without flag filter (trivial)
        * above one with early reg
      * early regs vs flags
 * import messages. 4h  #1382 
 * do the same process for MKii or CAii ANC. 4h
    * has the missed visit metric. 8h
 * ensure any special handling by middleware is reproduced; e.g. handling for extra columns

HR sum: 36h
MKii sum: 12h
total sum: 48h
probability hour estimate is anywhere near correct: same odds as winning the lotto

**Comment by browndav** [comment]
#1300 is split off from this.

**Comment by btbonval** [comment]
I'm giving serious consideration to creating a single form view which contains only the usual meta data (contact info, reported date) and nothing form-specific, except possibly the type of the form itself.

Then I create additional form views for each form type containing the form UUID as a pseudo-FK (no actual constraint) and the fields relevant to that form.

This is similar to what I have now, except it'd remove the usual meta data out of the formviews and into their own view. The additional form-generic view would be useful for counting various activity metrics and other stats that cross-cut forms. Presently I'm duplicating some viewform calculations and accessing the raw nosql data directly for the cross cutting calculations.

**Comment by btbonval** [comment]
New task that was missed from the list: accessibility to the calculations are needed. For now, we'll output postgres to Klipfolio. #1383 .

**Comment by btbonval** [comment]
I've managed to perform pretty much every calculation needed for impact and dashboards inside of postgres. It worked for HR, and then I was able to apply the same stuff to Baglung with little additional effort.

As far as this phase goes, I think the postgres middleware work is ready for active use. Decisions need to be made about when/how/where the adapter gets run and the database lives in production. Proof of concept looks great though.
